pkg/cafs/cafs.go

```diff
@@ -132,54 +122,85 @@ type cafsStore struct {
 }
 
 type defaultFs struct {
-	store                       cafsStore
-	leafSize                    uint32
-	prefix                      string
-	zl                          zap.Logger //nolint:structcheck,unused
-	l                           log.Logger //nolint:structcheck,unused
+	store    cafsStore
+	leafSize uint32
+	l        *zap.Logger
+
+	// prefix determines a namespace for keys
+	prefix string
+	pather func(Key) string // pathing logic
+
+	// buffer cache (atm only supported for ReadAt)
+	lru      *lru.Cache // this holds leaf data in cache
+	lruLatch sync.Mutex // this ensures consistent LRU buffer pinning
+	leafPool FreeList
+	lruSize  int
+
+	// root key cache of resolved leaf keys
+	keysCache     *lru.Cache // this holds leaf keys in cache to avoid resolving root keys again
+	keysCacheSize int
+
+	// options
 	leafTruncation              bool
-	lru                         *lru.Cache
-	leafPool                    *leafFreelist
 	concurrentFlushes           int
 	readerConcurrentChunkWrites int
 	deduplicationScheme         string
+	withPrefetch                int
+	withVerifyHash              bool
 }
 
 func (d *defaultFs) GetAddressingScheme() string {
 	return DeduplicationBlake
 }
 
 func (d *defaultFs) Put(ctx context.Context, src io.Reader) (PutRes, error) {
-	w := d.writer(d.prefix)
-	defer w.Close()
+	d.l.Debug("Start cafs Put")
+	defer d.l.Debug("End cafs Put")
+
+	w := d.writer()
+
+	// write leaf blobs
 	written, err := io.Copy(w, src)
 	if err != nil {
+		w.Close()
```

Comment:
nit: prefer `defer w.Close()` + function scope ... so

```
	w := d.writer()

	// write leaf blobs
	written, err := io.Copy(w, src)
	if err != nil {
		w.Close()
		return PutRes{}, err
	}
	root, keys, err := w.Flush()
	if err != nil {
		w.Close()
		return PutRes{}, err
	}

	if err = w.Close(); err != nil {
		return PutRes{}, err
	}
```

becomes

```
        var (
root Key
keys []Key
)

	// write leaf blobs
err := func() {
	w := d.writer()
defer w.Close()
	written, err := io.Copy(w, src)
	if err != nil {
		return err
	}
	root, keys, err = w.Flush()
	if err != nil {
		return err
	}
return nil
}
if err != nil { return PutRes{}, err }
```

these are not exactly equivalent, yet i think the second is more idiomatic?

like, don't we usually cleanup in golang with `defer <cleanup>`?

or is it that the writer here has a non-idiomatic `Close()` method?

Reply:
`defer` is indeed idiomatic. It is also slower (every go version comes with some improvements here, though).
But we also want to check the error returned by `w.Close` (which is important here, but only in the main path, not on other error handling).

Hence I specifically removed the `defer` to avoid closing twice, which could lead to unspecified results.

Reply:
follow-up note: disagree - won't fix

Reply:
np